### DISCUSSION TITLE
Workflow notes

### DISCUSSION THREAD

**Discussion by dodyg** [OrchardCMS/OrchardCore#15840] [Q&A]
I need to do some work that requires workflow so I am going to keep notes here.

**Comment by dodyg** [comment]
The problem with workflow task designer is that it's not very clear what is produced on each step. Outcomes has visual indication but there's no way to figure out what inside `Input`, `Output` or `Properties`.

For example, what will be the output of `Content Updated` event. Documentation will solve this issue but I think if there's a way for the designer to be able to read some metadata with these type of information it will be splendid.

**Comment by dodyg** [comment]
This seems important in figuring out the output/properties set by `Content Updated`.

https://github.com/OrchardCMS/OrchardCore/blob/12417a2629f3719a393d35789a5f9a45add8ee6d/src/OrchardCore.Modules/OrchardCore.Contents/Workflows/Handlers/ContentsHandler.cs

**Comment by dodyg** [comment]
> If the workflow was initiated by a content-related event (such as Published), then the content item will be available via the Content property, which you can then access from activities in your workflow (via both Liquid and JavaScript expressions). 

https://github.com/OrchardCMS/OrchardCore/issues/2240#issuecomment-414301635

So it should be available via ```{{ Workflow.Input.ContentItem.Content }}```

**Comment by jtkech** [comment]
I'm curently working on this, if your activity is directly after a content event e.g. `ContentCreatedEvent` it is available in `workflowContext.Input["ContentItem"]`, otherwise a previous activity may have set it in the `workflowContext.Properites["ContentItem"]` and / or `workflowContext.LastResult`.

As i remember e.g. through a javascript expression `input("ContentItem")`, didn't re-tried with a liquid expression

**Comment by dodyg** [comment]
I wish there's a Content shape viewer. Working with content with the workflow is quite cumbersome because you need to figure out the JSON structure. It's a trial and error process.

**Comment by dodyg** [comment]
The workflow is missing the ability to perform query. I will try to implement this.

**Comment by dodyg** [comment]
> I'm curently working on this, if your activity is directly after a content event e.g. ContentCreatedEvent it is available in workflowContext.Input["ContentItem"], otherwise a previous activity may have set it in the workflowContext.Properites["ContentItem"] and / or workflowContext.LastResult.

@jtkech 

`Workflow.Input.ContentItem` is available beyond immediate activity.

![not-immediate](https://user-images.githubusercontent.com/1761249/81044431-98616a80-8eb4-11ea-9c3b-f0c4a1b559a7.png)

**Comment by dodyg** [comment]
**Content Query Task**

This is a activity task that allows you to perform query, pass parameters and set the result into specific `Workflow.Properties`. It's very useful to check for content duplication, etc. Important when you need to decide whether to insert a new content or update an existing content.

![content-query-design](https://user-images.githubusercontent.com/1761249/81045163-ef1b7400-8eb5-11ea-92cb-523598558099.png)
![content-query-task](https://user-images.githubusercontent.com/1761249/81045164-f04ca100-8eb5-11ea-9e75-fbcf19a86a55.png)

This is the source code
``` csharp
  public class ContentQueryTask : TaskActivity
    {
        protected readonly IStringLocalizer S;
        private readonly IWorkflowExpressionEvaluator _expressionEvaluator;
        private readonly OrchardCore.IOrchardHelper _helper;

        public ContentQueryTask(
            IStringLocalizer<ContentQueryTask> localizer, 
            IWorkflowExpressionEvaluator expressionEvaluator,
            OrchardCore.IOrchardHelper helper)
        {
            S = localizer;
            _expressionEvaluator = expressionEvaluator;
            _helper = helper;
        }

        public override string Name => nameof(ContentQueryTask);

        public override LocalizedString DisplayText => S["Content Query Task"];

        public override LocalizedString Category => S["Content"];

        public override IEnumerable<Outcome> GetPossibleOutcomes(WorkflowExecutionContext workflowContext, ActivityContext activityContext)
        {
            return Outcomes(S["Done"]);
        }

        public override async Task<ActivityExecutionResult> ExecuteAsync(WorkflowExecutionContext workflowContext, ActivityContext activityContext)
        {
            var queryName = string.Empty;
            if (!string.IsNullOrWhiteSpace(QueryName.Expression))
                queryName = await _expressionEvaluator.EvaluateAsync(QueryName, workflowContext);

            var propertyName = "ContentQuery";
            if (!string.IsNullOrWhiteSpace(PropertyName.Expression))
                propertyName = await _expressionEvaluator.EvaluateAsync(PropertyName, workflowContext);

            var paramaters = "";
            if (!string.IsNullOrWhiteSpace(Parameters.Expression))
                paramaters = await _expressionEvaluator.EvaluateAsync(Parameters, workflowContext);

            IDictionary<string, object> @params = new Dictionary<string, object>();

            if (!string.IsNullOrWhiteSpace(paramaters))
                @params = JObject.Parse(paramaters).ToDictionary();

            var result = await _helper.ContentQueryAsync(queryName, @params);

            workflowContext.Properties[propertyName] = result;

            return Outcomes("Done");
        }

        public WorkflowExpression<string> QueryName
        {
            get => GetProperty(() => new WorkflowExpression<string>(""));
            set => SetProperty(value);
        }

        public WorkflowExpression<string> PropertyName
        {
            get => GetProperty(() => new WorkflowExpression<string>(""));
            set => SetProperty(value);
        }

        public WorkflowExpression<string> Parameters
        {
            get => GetProperty(() => new WorkflowExpression<string>(""));
            set => SetProperty(value);
        }
    }

    public class ContentQueryViewModel
    {
        public string QueryNameExpression { get; set; }

        public string PropertyNameExpression { get; set; }

        public string ParametersExpression { get; set; }
    }

    public class ContentQueryTaskDisplay: ActivityDisplayDriver<ContentQueryTask, ContentQueryViewModel>
    {
        protected override void EditActivity(ContentQueryTask activity, ContentQueryViewModel model)
        {
            model.QueryNameExpression = activity.QueryName.Expression;
            model.PropertyNameExpression = activity.PropertyName.Expression;
            model.ParametersExpression = activity.Parameters.Expression;
        }

        protected override void UpdateActivity(ContentQueryViewModel model, ContentQueryTask activity)
        {
            activity.QueryName = new WorkflowExpression<string>(model.QueryNameExpression);
            activity.PropertyName = new WorkflowExpression<string>(model.PropertyNameExpression);
            activity.Parameters = new WorkflowExpression<string>(model.ParametersExpression);
        }
    }
```

**Comment by dodyg** [comment]
**Content Query Task part 2**

You can see here how `ContentQueryTask` can be used as an input for an `IfTask`.

![content-query-task-2](https://user-images.githubusercontent.com/1761249/81046072-e166ee00-8eb7-11ea-91e6-da715a98818d.png)

**Comment by deanmarcussen** [comment]
> I wish there's a Content shape viewer. Working with content with the workflow is quite cumbersome because you need to figure out the JSON structure. It's a trial and error process.

See over here @dodyg https://github.com/OrchardCMS/OrchardCore/pull/5862 it should add a `View as JSON` button to the admin list, to make it a little easier to see the shape of what you're working with

**Comment by yash1529** [comment]
![image](https://user-images.githubusercontent.com/77006088/235135795-e77ee199-e3c3-4444-9b96-7b45c3038b30.png)

I want to access the highlighted field but i am unable to . I am only able to acceess CorrelationId,Input,output and properties. Can any1 please help me out with this?
If possible i want to access whole json object and forwaRD IF POSSIBLE? THANKS IN ADVANCE

**Comment by sebastienros** [comment]
@yash1529 you may have more answers by creating a new discussion, nobody will see this here.
